Add tutorial for HOHQMesh to Trixi docs (#791)
trixi-framework/Trixi.jl: docs/src/hohqmesh_tutorial.md

# Unstructured meshes with HOHQMesh.jl

Trixi.jl supports numerical approximations on unstructured quadrilateral meshes

> with the `UnstructuredMesh2D` mesh type.

Reviewer: ```suggestion
with the [`UnstructuredMesh2D`](@ref) mesh type.
```
This adds a link to the rendered docstring.

The purpose of this tutorial is to demonstrate how one can use the `UnstructuredMesh2D`
functionality of Trixi.jl. This begins by running and visualizing an available unstructured
quadrilateral mesh example. Then, the tutorial will demonstrate how one can
conceptualize a problem with curved boundaries, generate
a curvilinear mesh using the available software in the Trixi.jl ecosystem,
and then run a simulation using Trixi.jl on said mesh.

Unstructured quadrilateral meshes can be made

> with the [*High-Order Hex-Quad Mesh (HOHQMesh)*](https://github.com/trixi-framework/HOHQMesh)

Reviewer: What does the `*` in `(HOHQMesh)*` mean?

Author: I will remove this. It is supposed to make it italic but that was the style before I put in a link

> generator created and developed by David Kopriva.

Reviewer: Isn't "HOHQMesh" also the name of the "High-Order Hex-Quad Mesh generator"?

Author: I suppose when we all say "HOHQMesh" that we implicitly mean the generator without ever saying this extra word. Should I move generator into the brackets?

Reviewer: Yeah, that would be good :+1:

> HOHQMesh is a mesh generator specifically designed for spectral element methods where elements can be larger (due to the high accuracy

Reviewer: "larger" than what?

Author: compared to traditional finite element mesh generators. I can clarify this

Reviewer: That would be nice :+1:  Not everyone reading this will have our background and add these words automatically

of the spatial approximation) and provides high-order boundary curve information (needed to accurately set boundary conditions).
For more information about the design and features of HOHQMesh one can refer to its
[official documentation](https://trixi-framework.github.io/HOHQMesh/).

HOHQMesh is incorporated into the Trixi.jl framework via the registered Julia package
[HOHQMesh.jl](https://github.com/trixi-framework/HOHQMesh.jl).
This package provides a Julia wrapper for the HOHQMesh generator that allows users to easily create mesh files without the need to build
HOHQMesh from source. To install the HOHQMesh package execute
```julia
julia> import Pkg; Pkg.add("HOHQMesh")
```
Now we are ready to generate an unstructured quadrilateral mesh that can be used by Trixi.jl.

## Running and visualizing an unstructured simulation

Trixi.jl supports solving hyperbolic problems on several mesh types.
There is a default example for this mesh type that can be executed by
```julia

> julia> trixi_include(default_example_unstructured())

Reviewer: ```suggestion
julia> using Trixi; trixi_include(default_example_unstructured())
```

```
This will compute a smooth, manufactured solution test case for the 2D compressible Euler equations
on the curved quadrilateral mesh described in the
[Trixi documentation](https://trixi-framework.github.io/Trixi.jl/stable/meshes/unstructured_quad_mesh/).

Apart from the usual error and timing output provided by the Trixi.jl run, it is useful to visualize and inspect
the solution. Currently, for solutions on unstructured quadrilateral meshes, this requires post-processing the
Trixi.jl output file(s) using the [`Trixi2Vtk`](https://github.com/trixi-framework/Trixi2Vtk.jl) tool

> and plotting them with [ParaView](https://www.paraview.org/download/).

Reviewer: Since we will have support for other options soon, what about re-writing this such that Paraview is described as a possibility, not a necessity?

Author: That is true, e.g., now that the `Makie.jl` PRs are converging. I wasn't sure the best way to present this. Is it the case that Makie should supplant `Trixi2Vtk`? I.e, is the plan to deprecate `Trixi2Vtk` in the future?

Reviewer 2: No, IMHO Trixi2Vtk will never be deprecated, simply because of the ability to do visualization *and* postprocessing in ParaView directly.

Reviewer 2: Trixi2Vtk and visualization via Plots/Makie are two (three) tools with similar, but slightly different purposes.

Author: okay that is good to know. This is a product of my own confusion because I only use the ParaView approach

Reviewer: Trixi2Img is deprecated

Author: Okay, I updated the wording here to clarify that this is one option, not the only option, for unstructured mesh solutions

To convert the HDF5-formatted `.h5` output file(s) from Trixi into VTK format execute the following
```julia
julia> using Trixi2Vtk

julia> trixi2vtk("out/solution_000180.h5", output_directory="out")
```
Note this step takes about 15-30 seconds as the package `Trixi2Vtk` must be precompiled and executed for the first time
in your REPL session. The above `trixi2vtk` command will convert the solution file at the final time into a `.vtu` file

> which can be readin and visualize with ParaView. Optional arguments for `trixi2vtk` are: (1) Pointing to the `output_directory`

Reviewer: ```suggestion
which can be readin and visualized with ParaView. Optional arguments for `trixi2vtk` are: (1) Pointing to the `output_directory`
```

   number of iterations = 25
\end{SPRING_SMOOTHER}
```
Once HOHQMesh generates the mesh, a spring-mass-dashpot model is created to smooth the mesh and create "nicer" quadrilateral elements.
The [default parameters of Hooke's law](https://trixi-framework.github.io/HOHQMesh/the-control-input/#the-smoother)
for the spring-mass-dashpot model have been selected after a fair amount of experimentation across many meshes.
If you wish to deactivate this feature you can set `smoothing = OFF` (or remove this block from the control file).

After the `CONTROL_INPUT` environment block comes the `MODEL` environment block. It is here where the user can prescribe curved boundary information with either:
* An `OUTER_BOUNDARY` (covered in the next section of this tutorial).
* One or more `INNER_BOUNDARIES`.

There are several options to describe the boundary curve data to HOHQMesh like splines or parametric curves.

For the example `ice_cream_straight_sides.control` we define three internal boundaries; two straight-sided and
one as a circular arc.
Within the HOHQMesh control input each curve must be assigned to a `CHAIN` as shown below in the complete

> `INNER_BOUNDARIES` block

Reviewer: ```suggestion
`INNER_BOUNDARIES` block.
```

      name = RightSlant
      xStart = [ 0.0, -3.0, 0.0]
      xEnd   = [ 2.0, 1.0, 0.0]
    \end{END_POINTS_LINE}

    \begin{CIRCULAR_ARC}
      name        = IceCream
      units       = degrees
      center      = [ 0.0, 1.0, 0.0]
      radius      = 2.0
      start angle = 0.0
      end angle   = 180.0
    \end{CIRCULAR_ARC}
  \end{CHAIN}

\end{INNER_BOUNDARIES}
```
It is important to note there are two `name` quantities one for the `CHAIN` and one for the `PARAMETRIC_EQUATION_CURVE`.
The name for the `CHAIN` is used internally by HOHQMesh, so if you have multiple `CHAIN`s they **must be given a unique name**.
The name for the `PARAMETRIC_EQUATION_CURVE` with be printed to the appropriate boundaries within the `.mesh` file produced by
HOHQMesh. The resulting mesh generated by HOHQMesh.jl is given in the following figure.
![mesh_straight_sides](https://user-images.githubusercontent.com/25242486/129603958-08e4b874-53d5-4511-9a54-6daf4c21edca.png)
The mesh file `ice_cream_straight_sides.mesh` and its associated file for plotting
`ice_cream_straight_sides.tec` are placed in the `out` folder.
We note that Trixi.jl uses the boundary name information from the control file

> to assign boundary conditions in an elixir file.

Author: I like this emphasis!

  # set the freestream flow parameters
  rho_freestream = 1.0
  u_freestream = 0.3
  p_freestream = inv(equations.gamma)

  theta = 0.0 # zero angle of attack
  si, co = sincos(theta)
  v1 = u_freestream * co
  v2 = u_freestream * si

  prim = SVector(rho_freestream, v1, v2, p_freestream)
  return prim2cons(prim, equations)
end

# initial condition
initial_condition = uniform_flow_state

# boundary condition types
boundary_condition_uniform_flow = BoundaryConditionDirichlet(uniform_flow_state)
boundary_condition_slip_wall = BoundaryConditionWall(boundary_state_slip_wall)

# boundary condition dictionary
boundary_conditions = Dict( :Bottom     => boundary_condition_uniform_flow,
                            :Top        => boundary_condition_uniform_flow,
                            :Right      => boundary_condition_uniform_flow,
                            :Left       => boundary_condition_uniform_flow,
                            :LeftSlant  => boundary_condition_slip_wall,
                            :RightSlant => boundary_condition_slip_wall,
                            :IceCream   => boundary_condition_slip_wall )

# DGSEM solver.

> #    1) polydeg must be >= the polynomial order set in the HOHQMesh control file to guarantee

Reviewer 2: Isn't it 
```suggestion
#    1) polydeg must be >= two times the polynomial order set in the HOHQMesh control file to guarantee
```
or was that just for certain mesh types (with hanging nodes?)?

Author: Sort of, this is related to 3D curvilinear meshes where there are stronger constraints on the mapping. For 2D curvilinear `polydeg >= polydeg_mesh` is sufficient.

Reviewer 2: Thanks for the clarification!

summary_callback = SummaryCallback()
save_solution = SaveSolutionCallback(interval=10,
                                     save_initial_solution=true,
                                     save_final_solution=true)
stepsize_callback = StepsizeCallback(cfl=1.0)

callbacks = CallbackSet(summary_callback, save_solution, stepsize_callback)

# Evolve ODE problem in time using `solve` from OrdinaryDiffEq
sol = solve(ode, CarpenterKennedy2N54(williamson_condition=false),
            dt=1.0, # solve needs some value here but it will be overwritten by the stepsize_callback
            save_everystep=false, callback=callbacks);
# print the timer summary
summary_callback()
```

Visualization of the solution is carried out in a similar way as above. That is, one converts the `.h5`
output files with `trixi2vtk` and then plot the solution in ParaView. An example plot of the pressure
at the final time is shown below.

![simulation_straight_sides](https://user-images.githubusercontent.com/25242486/129733926-6ef80676-779b-4f1e-9826-3ebf750cf382.png)

## Making a mesh with a curved outer boundary

Let us modify the mesh from the previous task and place a circular outer boundary instead

> of straight sided outer boundaries.

Reviewer 2: ```suggestion
of straight-sided outer boundaries.
```

          name = OuterCircle
          xEqn = x(t) = 8.0*sin(2.0*pi*t)
          yEqn = y(t) = 8.0*cos(2.0*pi*t)
          zEqn = z(t) = 0.0
       \end{PARAMETRIC_EQUATION_CURVE}

    \end{OUTER_BOUNDARY}

    \begin{INNER_BOUNDARIES}

       \begin{CHAIN}
         name = IceCreamCone
         \begin{END_POINTS_LINE}
            name = LeftSlant
            xStart = [-2.0, 1.0, 0.0]
            xEnd   = [ 0.0, -3.0, 0.0]
         \end{END_POINTS_LINE}

         \begin{END_POINTS_LINE}
            name = RightSlant
            xStart = [ 0.0, -3.0, 0.0]
            xEnd   = [ 2.0, 1.0, 0.0]
         \end{END_POINTS_LINE}

         \begin{CIRCULAR_ARC}
            name        = IceCream
            units       = degrees
            center      = [ 0.0, 1.0, 0.0]
            radius      = 2.0
            start angle = 0.0
            end angle   = 180.0
         \end{CIRCULAR_ARC}
       \end{CHAIN}

    \end{INNER_BOUNDARIES}

\end{MODEL}
\end{FILE}
```

The first alteration is that we have altered the second block of information
`BACKGOUND_GRID` within the `CONTROL_INPUT` to be
```
\begin{BACKGROUND_GRID}
   background grid size = [1.0, 1.0, 0.0]
\end{BACKGROUND_GRID}
```

> Because this mesh control file has an outer boundary that determines the extent of the domain to be meshed.
> Therefore, we only need to supply the `background grid size` to the `BACKGROUND_GRID` control input.

Reviewer 2: ```suggestion
This mesh control file has an outer boundary that determines the extent of the domain to be meshed.
Therefore, we only need to supply the `background grid size` to the `BACKGROUND_GRID` control input.
```

In this case it is a circle of radius `8` centered at `[0.0, 0.0, 0.0]` written as a set of
`PARAMETRIC_EQUATION_CURVE`s.
```
    \begin{OUTER_BOUNDARY}

       \begin{PARAMETRIC_EQUATION_CURVE}
          name = OuterCircle
          xEqn = x(t) = 8.0*sin(2.0*pi*t)
          yEqn = y(t) = 8.0*cos(2.0*pi*t)
          zEqn = z(t) = 0.0
       \end{PARAMETRIC_EQUATION_CURVE}

    \end{OUTER_BOUNDARY}
```
Just as with the inner boundary curves, we must assign a name to the `OUTER_BOUNDARY`. It will be included
in the generated `.mesh` file and is used within the Trixi.jl elixir file to set boundary conditions.

The resulting mesh generated by HOHQMesh.jl is given in the following figure.

![mesh_curved_sides](https://user-images.githubusercontent.com/25242486/129603957-6a92618f-9ed8-4072-b6ab-05533bea746a.png)

Again, the mesh file `ice_cream_curved_sides.mesh` and its associated file for plotting
`ice_cream_curved_sides.tec` are placed in the `out` folder.

## Running Trixi.jl on `ice_cream_curved_sides.mesh`

We can reuse much of the elixir file to setup the uniform flow over an ice cream cone from the

> previous part of this tutorial. The only components of the elixir file that must be changed are the boundary condition
> dictionary because we now have a boundary named `OuterCirle` instead of four edges of a bounding box.

Reviewer 2: ```suggestion
previous part of this tutorial. The only component of the elixir file that must be changed is the boundary condition
dictionary because we now have a boundary named `OuterCirle` instead of four edges of a bounding box.
```
